### ISSUE TITLE
Bidirectional Module Federation Sharing is Not Working – Browser Hangs on Startup with Numerous Warnings

### ISSUE THREAD

**Issue by Emehado** [module-federation/vite#180]
There is a problem with bidirectional module federation between micro frontends. Upon loading the application, the browser hangs and is unresponsive, with numerous warnings in the console. 

## Steps to Reproduce
- clone https://github.com/Emehado/react-microfrontend-demo/tree/issue/bidirectional-hangs-on-startup
- Start the application.
- Open the application in the browser.

<img width="492" alt="Screenshot 2024-11-06 at 11 42 52 AM" src="https://github.com/user-attachments/assets/832d8e5b-f71e-41d5-9542-cd681705e3e5">

**Comment by Emehado** [comment]
Thanks for the quick response @zhangHongEn but just tested with v1.1.3 and it also has same problem.

**Comment by zhangHongEn** [comment]
I tried it, but everything works fine.

**Comment by gioboa** [comment]
@Emehado can you share with us your system info please?
```
npx envinfo --system --binaries --browsers
```

**Comment by Emehado** [comment]
@gioboa here you go
```
System:
    OS: macOS 15.1
    CPU: (10) x64 Apple M1 Pro
    Memory: 23.71 MB / 16.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 20.17.0 - ~/.volta/tools/image/node/20.17.0/bin/node
    Yarn: 1.22.18 - ~/.volta/tools/image/yarn/1.22.18/bin/yarn
    npm: 10.8.2 - ~/.volta/tools/image/node/20.17.0/bin/npm
    pnpm: 9.12.3 - ~/.volta/bin/pnpm
    Watchman: 2024.10.07.00 - /opt/homebrew/bin/watchman
  Browsers:
    Brave Browser: 130.1.71.121
    Chrome: 130.0.6723.117
    Edge: 130.0.2849.80
    Safari: 18.1

```

**Comment by gioboa** [comment]
@Emehado You are doing a recursive scenario :thinking: 
`host -> remote -> host --> remote -> host [...]`
What's the purpose of this?

**Comment by Emehado** [comment]
@gioboa, The purpose is to be able to share code vice-versa. If bi-directional code sharing doesn't work, it means you'd need to have a third MFE which can share code to your host and remotes. If Bi-directional works fine, then you could expose modules from your host to be used in remotes. (eg. sharing color theme state). Also, I'm not sure what you mean by recursive. I found this concept which is supposed to work fine and defined as Bi-directional. Similar case is described [here](https://github.com/module-federation/module-federation-examples/blob/master/bi-directional/README.md).

Is this not correct behavior?

**Comment by gioboa** [comment]
The problem is that Remote app is loading the Host 
![image](https://github.com/user-attachments/assets/d4b8c5d2-2e54-4693-92fc-3a87077ab397)
if you need to share things you can look at this [Vue example](https://github.com/gioboa/vue-microfrontend-demo/blob/2b5ea62c332c0e8ed1605ba242346c65760e6554/remote/vite.config.ts#L47) I did

**Comment by Emehado** [comment]
That's just a bad example. I can edit the example repo to only load some component from the host instead of the entire host and the problem would still persist.

Also I see your example but wouldn't that also require a rebuild of all dependents? if yes, then that might defeat purpose of runtime code sharing.

**Comment by gioboa** [comment]
In my example you can share things from mfes without rebuilding anything. I'm looking forward to seeing the new example.

**Comment by zhangHongEn** [comment]
@Emehado I switched to your branch and did reproduce the issue. Have you tried using Webpack or Rspack?

**Comment by zhangHongEn** [comment]
This feature can be implemented, but it requires ensuring that @module-federation/core supports it.

**Comment by Emehado** [comment]
@zhangHongEn I tried using rsbuild which supports this feature but HMR I get from rsbuild was not great (4s) and when I add module federation, it goes nuts. Maybe I'm doing somethig wrong but Vite on the other hand is its still almost instantaneous. That's why I have resolved to use vite thus far.

**Comment by Emehado** [comment]
@gioboa , I have updated the example branch to only consume a component from the host instead of the entire host.
https://github.com/Emehado/react-microfrontend-demo/tree/issue/bidirectional-hangs-on-startup.

**Comment by ScriptedAlchemy** [comment]
> @zhangHongEn I tried using rsbuild which supports this feature but HMR I get from rsbuild was not great (4s) and when I add module federation, it goes nuts. Maybe I'm doing somethig wrong but Vite on the other hand is its still almost instantaneous. That's why I have resolved to use vite thus far.

Turn off typescript dts stuff. Our remote type generation accounts for 95% of the slowdown. 

Dts:false on the plugin and hmr will take 50ms or less.

**Comment by zhangHongEn** [comment]
@ScriptedAlchemy Does rspack support circular references where the host imports a remote, and the remote imports the host?

**Comment by ScriptedAlchemy** [comment]
Yes. Nested circular and multi cyclic as well as host uses own remote or remote uses hosts remote patterns are all supported

**Comment by Emehado** [comment]
Thanks @ScriptedAlchemy, But without module federation, HMR was already much slower using rsbuild. Any other pointers that could be causing that? Testing with a fresh project was lightning fast, I would really love to push through with rsbuild if I could get HMR to vite-like speed.

**Comment by ScriptedAlchemy** [comment]
Do you have a repo or zoom call so I can discuss? Dm me on Twitter. 

Bytedance has some of the world's largest apps and the slowest hmr on record was 1s, the repo is 40gb

**Comment by danielBence** [comment]
I recon this could be closed?  1.1.7 fixed this bug for me atleast thanks to @zhangHongEn   :)

**Comment by gioboa** [comment]
Yep, thanks @danielBence
